### ISSUE TITLE
Error 500 for multiple files of the documentation of a package

### ISSUE THREAD

**Issue by jrast** [devpi/devpi#846]
After I replaced a package in a volatile index, I noticed that the documentation was no longer correctly serverd. A look into the logfiles showed a lot of exceptions:

````
2021-08-22 00:06:54,185 INFO  [req12149] [Wtx716] set_metadata robdevsim-0.2.1
2021-08-22 00:06:54,310 INFO  [req12149] [Wtx716] fswriter717: commited at 717
2021-08-22 00:06:54,732 INFO  [NOTI] [Rtx717] Processed a total of 1 projects and queued 1
2021-08-22 00:06:55,310 INFO  [IDX] Indexer queue size ~ 1
2021-08-22 00:06:55,310 INFO  [NOTI] [Rtx717] Processed a total of 1 projects and queued 1
2021-08-22 00:06:55,435 ERROR [NOTI] Unhandled exception in notification thread.
Traceback (most recent call last):
  File "d:\python\lib\site-packages\devpi_server\fileutil.py", line 14, in rename
    os.rename(source, dest)
FileExistsError: [WinError 183] Cannot create a file when that file already exists: 'D:\\Data\\devpi\\.event_serial-tmp' -> 'D:\\Data\\devpi\\.event_serial'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\python\lib\site-packages\devpi_server\keyfs.py", line 107, in thread_run
    self.tick()
  File "d:\python\lib\site-packages\devpi_server\keyfs.py", line 94, in tick
    self.write_event_serial(event_serial)
  File "d:\python\lib\site-packages\devpi_server\keyfs.py", line 82, in write_event_serial
    write_int_to_file(event_serial + 1, self.event_serial_path)
  File "d:\python\lib\site-packages\devpi_server\fileutil.py", line 114, in write_int_to_file
    rename(tmp_path, path)
  File "d:\python\lib\site-packages\devpi_server\fileutil.py", line 20, in rename
    os.remove(dest)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'D:\\Data\\devpi\\.event_serial'
2021-08-22 00:06:57,357 INFO  [NOTI] [Rtx717] Processed a total of 1 projects and queued 1
2021-08-22 00:06:57,497 ERROR [NOTI] calling <bound method EventSubscribers.on_changed_file_entry of <devpi_server.model.EventSubscribers object at 0x0000025BCD1E3DD8>> failed, serial=717
Traceback (most recent call last):
  File "d:\python\lib\site-packages\py\_error.py", line 83, in checked_call
    cls = self._geterrnoclass(_winerrnomap[errno])
KeyError: 41

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\python\lib\site-packages\devpi_web\doczip.py", line 40, in locked_unpack_path
    yield (hash_file, unpack_path)
  File "d:\python\lib\site-packages\devpi_web\doczip.py", line 165, in remove_docs
    directory.remove()
  File "d:\python\lib\site-packages\py\_path\local.py", line 217, in remove
    ignore_errors=ignore_errors)
  File "d:\python\lib\site-packages\py\_error.py", line 85, in checked_call
    raise value
  File "d:\python\lib\site-packages\py\_error.py", line 66, in checked_call
    return func(*args, **kwargs)
  File "d:\python\lib\shutil.py", line 494, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "d:\python\lib\shutil.py", line 384, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "d:\python\lib\shutil.py", line 384, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "d:\python\lib\shutil.py", line 384, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "d:\python\lib\shutil.py", line 393, in _rmtree_unsafe
    onerror(os.rmdir, path, sys.exc_info())
  File "d:\python\lib\shutil.py", line 391, in _rmtree_unsafe
    os.rmdir(path)
OSError: [WinError 145] The directory is not empty: 'D:\\Data\\devpi\\hamilton\\staging\\robdevsim\\0.2.1\\+doc\\_static\\css\\fonts'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\python\lib\site-packages\py\_error.py", line 66, in checked_call
    return func(*args, **kwargs)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'D:\\Data\\devpi\\hamilton\\staging\\robdevsim\\0.2.1\\+doc.hash'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\python\lib\site-packages\devpi_server\keyfs.py", line 206, in _execute_hooks
    sub(ev)
  File "d:\python\lib\site-packages\devpi_server\model.py", line 1700, in on_changed_file_entry
    relpath=ev.typedkey.relpath
  File "d:\python\lib\site-packages\pluggy\hooks.py", line 289, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "d:\python\lib\site-packages\pluggy\manager.py", line 68, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "d:\python\lib\site-packages\pluggy\manager.py", line 62, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "d:\python\lib\site-packages\pluggy\callers.py", line 208, in _multicall
    return outcome.get_result()
  File "d:\python\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "d:\python\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "d:\python\lib\site-packages\devpi_web\main.py", line 340, in devpiserver_on_remove_file
    remove_docs(stage, project, version)
  File "d:\python\lib\site-packages\devpi_web\doczip.py", line 165, in remove_docs
    directory.remove()
  File "d:\python\lib\contextlib.py", line 99, in __exit__
    self.gen.throw(type, value, traceback)
  File "d:\python\lib\site-packages\devpi_web\doczip.py", line 47, in locked_unpack_path
    hash_path.remove()
  File "d:\python\lib\site-packages\py\_path\local.py", line 223, in remove
    py.error.checked_call(os.remove, self.strpath)
  File "d:\python\lib\site-packages\py\_error.py", line 86, in checked_call
    raise cls("%s%r" % (func.__name__, args))
py.error.EBUSY: [Resource device]: remove('D:\\Data\\devpi\\hamilton\\staging\\robdevsim\\0.2.1\\+doc.hash',)
2021-08-22 00:06:57,950 INFO  [IDX] Committing 18 new documents to search index.
```` 

As you can see, the server is running on a windows system and it seems like there are some permission issues as on window a file can not be deleted while accessed. 

**Versions:**
OS: Windows 10
Installed Packages:
````
apipkg==1.5
appdirs==1.4.3
argon2-cffi==19.1.0
attrs==20.3.0
beautifulsoup4==4.7.1
bleach==3.1.0
certifi==2019.3.9
cffi==1.12.2
Chameleon==3.6
chardet==3.0.4
check-manifest==0.37
cmarkgfm==0.4.2
defusedxml==0.5.0
devpi-client==5.2.2
devpi-common==3.6.0
devpi-server==6.2.0
devpi-web==4.0.8
docutils==0.14
execnet==1.5.0
filelock==3.0.10
hupper==1.6.1
idna==2.8
itsdangerous==1.1.0
lazy==1.4
passlib==1.7.1
PasteDeploy==2.0.1
pkginfo==1.5.0.1
plaster==1.0
plaster-pastedeploy==0.6
pluggy==0.9.0
py==1.8.0
pycparser==2.19
Pygments==2.3.1
pyramid==2.0
pyramid-chameleon==0.3
python-dateutil==2.8.0
readme-renderer==24.0
repoze.lru==0.7
requests==2.21.0
ruamel.yaml==0.15.89
six==1.12.0
soupsieve==1.8
strictyaml==1.0.0
toml==0.10.0
tox==3.7.0
translationstring==1.3
urllib3==1.24.1
venusian==1.2.0
virtualenv==16.0.0
waitress==1.2.1
webencodings==0.5.1
WebOb==1.8.5
Whoosh==2.7.4
zope.deprecation==4.4.0
zope.interface==4.6.0
````

**Comment by fschulze** [comment]
Did you get errors about the ``.event_serial`` file before you noticed these problems?

Is there anything else accessing the documentation files besides devpi-server? The file locking used should prevent these cases, but it only works if devpi-server is the only process doing so. Do you have nginx in front which serves the documentation directly and could interfere?

**Comment by jrast** [comment]
I did not watch out for errors about the .event_serial before, so I can't tell if this happend before. It took us a while before noticing this problem as the documentation is not accessed regulary on our staging index.

We don't use nginx or another server in front of the devpi server, so there should be no other processes which access the files. However, this is running on a Windows Server (Windows 2016) and Windows Defender Service is running, so the Virus Scann might access the files. Also the Server is running in a VM and I don't know exactly how the storage is attached, so there might be some delays on file access.

I tried to update devpi-server and devpi-web (already updated in the pip freeze output above) but this did not solve the problem. Are there other dependencies which I should update? Like waitress, pyramid, ... ?

**Comment by fschulze** [comment]
As long as the storage isn't seen as a network filesystem it should be fine. Any NFS or Samba/Windows Share is a no go, as the file locking doesn't work properly there. The storage can be on the network if it shows up as a regular file system (NTFS) inside the VM.

Is the problem persisting, or does it clear up after a while? If it persists, you can try to remove ``D:\\Data\\devpi\\hamilton\\staging\\robdevsim\\0.2.1\\`` manually. It will be recreated on demand.

Also it seems weird that ``D:\\Data\\devpi\\hamilton\\staging\\robdevsim\\0.2.1\\+doc\\_static\\css\\fonts`` isn't empty. Could it be that it contains something which can't be removed due to permissions?

**Comment by jrast** [comment]
The storage (D: Drive) is shown as "VMWare Virtual Disk SCSI Disk Device" so there should be no issue if I understand correctly. 

The problem is persisting, also uploading a new version (devpi upload --with-docs) does not recover from the wrong behaving state. Removing the directory manually does solve the problem.

**Comment by jrast** [comment]
Seems like the package update did not solve all the problems. We still get some errors:

````
2021-09-21 09:04:22,420 INFO  [req104198] POST /+login
2021-09-21 09:04:23,380 INFO  [req104199] POST /hamilton/staging/
2021-09-21 09:04:23,380 INFO  [req104199] [Wtx2428] set_metadata hamutils-0.7.0
2021-09-21 09:04:23,380 INFO  [req104199] hamilton/staging: got submit release info 'hamutils'
2021-09-21 09:04:23,458 INFO  [req104199] [Wtx2428] fswriter2429: commited at 2429
2021-09-21 09:04:23,474 INFO  [req104200] POST /hamilton/staging/
2021-09-21 09:04:23,474 INFO  [req104200] [Wtx2429] set_metadata hamutils-0.7.0
2021-09-21 09:04:23,490 INFO  [req104200] hamilton/staging: got submit release info 'hamutils'
2021-09-21 09:04:23,490 INFO  [req104200] [Wtx2429] added 'releasefile' link hamilton/staging/+f/721/6e3b286ed7b9a/hamutils-0.7.0-py3-none-any.whl
2021-09-21 09:04:23,490 INFO  [req104200] [Wtx2429] set_metadata hamutils-0.7.0
2021-09-21 09:04:23,521 INFO  [req104200] [Wtx2429] fswriter2430: commited at 2430
2021-09-21 09:04:23,536 INFO  [req104201] POST /hamilton/staging/
2021-09-21 09:04:23,536 INFO  [req104201] [Wtx2430] set_metadata hamutils-0.7.0
2021-09-21 09:04:23,536 INFO  [req104201] hamilton/staging: got submit release info 'hamutils'
2021-09-21 09:04:23,568 INFO  [req104201] [Wtx2430] fswriter2431: commited at 2431
2021-09-21 09:04:23,615 INFO  [req104202] POST /hamilton/staging/
2021-09-21 09:04:23,911 INFO  [req104202] [Wtx2431] set_metadata hamutils-0.7.0
2021-09-21 09:04:24,099 INFO  [req104202] hamilton/staging: got submit release info 'hamutils'
2021-09-21 09:04:24,115 INFO  [req104202] [Wtx2431] added 'releasefile' link hamilton/staging/+f/61d/f099c304c3a78/hamutils-0.7.0.tar.gz
2021-09-21 09:04:24,115 INFO  [req104202] [Wtx2431] set_metadata hamutils-0.7.0
2021-09-21 09:04:24,115 INFO  [NOTI] [Rtx2429] Processed a total of 1 projects and queued 1
2021-09-21 09:04:24,255 INFO  [req104202] [Wtx2431] fswriter2432: commited at 2432
2021-09-21 09:04:24,833 INFO  [NOTI] [Rtx2430] Processed a total of 1 projects and queued 1
2021-09-21 09:04:25,880 INFO  [NOTI] [Rtx2431] Processed a total of 1 projects and queued 1
2021-09-21 09:04:26,021 INFO  [req104203] POST /hamilton/staging/
2021-09-21 09:04:26,161 INFO  [req104203] [Wtx2432] set_metadata hamutils-0.6.2
2021-09-21 09:04:26,161 INFO  [req104203] hamilton/staging: got submit release info 'hamutils'
2021-09-21 09:04:26,193 INFO  [req104203] [Wtx2432] deleted 'doczip' link hamilton/staging/+f/a87/1c28918d4d002/hamutils-0.6.2.doc.zip
2021-09-21 09:04:26,193 INFO  [req104203] [Wtx2432] set_metadata hamutils-0.6.2
2021-09-21 09:04:26,271 INFO  [req104203] [Wtx2432] added 'doczip' link hamilton/staging/+f/a87/1c28918d4d002/hamutils-0.6.2.doc.zip
2021-09-21 09:04:26,271 INFO  [req104203] [Wtx2432] set_metadata hamutils-0.6.2
2021-09-21 09:04:26,474 ERROR Exception while serving /hamilton/staging/
Traceback (most recent call last):
  File "d:\python\lib\site-packages\devpi_server\fileutil.py", line 14, in rename
    os.rename(source, dest)
FileExistsError: [WinError 183] Cannot create a file when that file already exists: 'D:\\Data\\devpi\\+files\\hamilton\\staging\\+f\\a87\\1c28918d4d002\\hamutils-0.6.2.doc.zip-edf0710d-tmp' -> 'D:\\Data\\devpi\\+files\\hamilton\\staging\\+f\\a87\\1c28918d4d002\\hamutils-0.6.2.doc.zip'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\python\lib\site-packages\waitress\channel.py", line 397, in service
    task.service()
  File "d:\python\lib\site-packages\waitress\task.py", line 168, in service
    self.execute()
  File "d:\python\lib\site-packages\waitress\task.py", line 434, in execute
    app_iter = self.channel.server.application(environ, start_response)
  File "d:\python\lib\site-packages\devpi_server\views.py", line 166, in __call__
    return self.app(environ, start_response)
  File "d:\python\lib\site-packages\pyramid\router.py", line 270, in __call__
    response = self.execution_policy(environ, self)
  File "d:\python\lib\site-packages\pyramid\router.py", line 276, in default_execution_policy
    return router.invoke_request(request)
  File "d:\python\lib\site-packages\pyramid\router.py", line 245, in invoke_request
    response = handle_request(request)
  File "d:\python\lib\site-packages\devpi_server\views.py", line 190, in request_log_handler
    response = handler(request)
  File "d:\python\lib\site-packages\devpi_server\views.py", line 226, in request_tx_handler
    response = handler(request)
  File "d:\python\lib\contextlib.py", line 88, in __exit__
    next(self.gen)
  File "d:\python\lib\site-packages\devpi_server\keyfs.py", line 399, in transaction
    self.commit_transaction_in_thread()
  File "d:\python\lib\site-packages\devpi_server\keyfs.py", line 387, in commit_transaction_in_thread
    self._threadlocal.tx.commit()
  File "d:\python\lib\site-packages\devpi_server\keyfs.py", line 721, in commit
    commit_serial = self.conn.last_changelog_serial + 1
  File "d:\python\lib\site-packages\devpi_server\keyfs_sqlite_fs.py", line 207, in __exit__
    self.commit_to_filesystem(pending_renames)
  File "d:\python\lib\site-packages\devpi_server\keyfs_sqlite_fs.py", line 240, in commit_to_filesystem
    files_commit, files_del = commit_renames(basedir, rel_renames)
  File "d:\python\lib\site-packages\devpi_server\keyfs_sqlite_fs.py", line 306, in commit_renames
    rename(path, path[:-suffix_len])
  File "d:\python\lib\site-packages\devpi_server\fileutil.py", line 20, in rename
    os.remove(dest)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'D:\\Data\\devpi\\+files\\hamilton\\staging\\+f\\a87\\1c28918d4d002\\hamutils-0.6.2.doc.zip'
````

**Comment by fschulze** [comment]
I guess we can use this instead of ``rename``:
https://docs.python.org/3/library/os.html?highlight=replace#os.replace
Also see:
https://bugs.python.org/issue8828

**Comment by jrast** [comment]
You mean in devpi_server.fileutil?

````
def rename(source, dest):
    try:
        os.rename(source, dest)  # First this fails
    except OSError:
        destdir = os.path.dirname(dest)
        if not os.path.exists(destdir):
            os.makedirs(destdir)
        if sys.platform == "win32" and os.path.exists(dest):
            os.remove(dest)   # This line fails also! How to find out which process accesses the file?
        os.rename(source, dest)
````

**Comment by fschulze** [comment]
Yes, but that only fixes one of the issues. I still don't understand why the PermissionError happens with ``The process cannot access the file because it is being used by another process``.

**Comment by fschulze** [comment]
This should be fixed with devpi-server 6.5.0. If these errors persist, please create a new ticket with current tracebacks.
